### DISCUSSION TITLE
stdout synchronization

### DISCUSSION THREAD

**Discussion by officer-tuba** [gabime/spdlog#3151] [Q&A] [answered]
I'm working on a program that has both stdout logging through spdlog and a built-in interactive "terminal" for program commands.
i was able to greatly reduce races on stdout by inheriting from `stdout_color_sink_mt` and overriding the `log` function. but there's still an occasional "overlap" (an output line has a mix of spdlog output and my output).
i'm happy with the results as they are, but it feels like a small loose end.
is there a better way to sync with spdlog's stdout operations?

![log](https://github.com/user-attachments/assets/ba5ab08a-0229-44bc-97c2-f6e7d1836992)

**Comment by tt4g** [comment]
Avoiding mixed prints to stdout can be achieved with critical sections in a single process.
However, it is difficult with multiple processes.
This is because it is not possible to perform global locking on the OS buffering function.

You might be able to achieve your goal by piping the output of multiple processes, but my guess is that it would be quite difficult.

See also: https://unix.stackexchange.com/a/476089

**Reply by tt4g** [reply]
It is a less than perfect workaround, but by calling `logger->flush_on(spdlog::level::trace);`, spdlog will actively flush stdout.
As long as the OS buffer size is not exceeded, this may avoid the problem.

**Reply by officer-tuba** [reply]
> Avoiding mixed prints to stdout can be achieved with critical sections in a single process. However, it is difficult with multiple processes. This is because it is not possible to perform global locking on the OS buffering function.
> 
> You might be able to achieve your goal by piping the output of multiple processes, but my guess is that it would be quite difficult.
> 
> See also: https://unix.stackexchange.com/a/476089


it's a single process, so i don't think this applies.

the gif in my original post shows the program running - i'm typing into the prompt at the bottom as log messages scroll by. it works really well about 99% of the time. occasionally a log output line will interleave with the prompt input.
i'm aware of critical sections, just not how to cooperate with spdlog's stdout printing. which is essentially my original question.

**Reply by officer-tuba** [reply]
> It is a less than perfect workaround, but by calling `logger->flush_on(spdlog::level::trace);`, spdlog will actively flush stdout. As long as the OS buffer size is not exceeded, this may avoid the problem.

i tried this, but i'm still seeing the occasional interleaving.

**Reply by tt4g** [reply]
> it's a single process, so i don't think this applies.
> 
> the gif in my original post shows the program running - i'm typing into the prompt at the bottom as log messages scroll by. it works really well about 99% of the time. occasionally a log output line will interleave with the prompt input.
> i'm aware of critical sections, just not how to cooperate with spdlog's stdout printing. which is essentially my original question.

That is, two processes, your application and the terminal, are printing to stdout.
The spdlog cannot mutex prevent different processes from accessing stdout.

**Reply by officer-tuba** [reply]
> > it's a single process, so i don't think this applies.
> > the gif in my original post shows the program running - i'm typing into the prompt at the bottom as log messages scroll by. it works really well about 99% of the time. occasionally a log output line will interleave with the prompt input.
> > i'm aware of critical sections, just not how to cooperate with spdlog's stdout printing. which is essentially my original question.
> 
> That is, two processes, your application and the terminal, are printing to stdout. The spdlog cannot mutex prevent different processes from accessing stdout.

thanks, i appreciate your input. i think we’re talking about two different things though.

**Reply by tt4g** [reply, marked as answer]
Ah, I finally understand.
So you have one multi-threaded application with the terminal running in a dedicated thread (and spdlog in a another thread).
Sorry it took me so long to understand.

Then use the mutex that spdlog uses internally to protect access to stdout.
You can access it using `spdlog::details::console_mutex::mutex()`:

https://github.com/gabime/spdlog/blob/5ebfc927306fd7ce551fa22244be801cf2b9fdd9/include/spdlog/details/console_globals.h#L12-L18

https://github.com/gabime/spdlog/blob/5ebfc927306fd7ce551fa22244be801cf2b9fdd9/include/spdlog/sinks/stdout_sinks.h#L59-L63

**Reply by officer-tuba** [reply]
> Ah, I finally understand. So you have one multi-threaded application with the terminal running in a dedicated thread (and spdlog in a another thread). Sorry it took me so long to understand.
> 
> Then use the mutex that spdlog uses internally to protect access to stdout. You can access it using `spdlog::details::console_mutex::mutex()`:
> 
> https://github.com/gabime/spdlog/blob/5ebfc927306fd7ce551fa22244be801cf2b9fdd9/include/spdlog/details/console_globals.h#L12-L18
> 
> https://github.com/gabime/spdlog/blob/5ebfc927306fd7ce551fa22244be801cf2b9fdd9/include/spdlog/sinks/stdout_sinks.h#L59-L63

that’s exactly what i was hoping to find. thank you so much!
sorry i wasn’t clear in my original post.
